src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -0,0 +1,416 @@
+	StringInfo ownerResetCommand = makeStringInfo();
+	char *qualifiedRelationName = generate_qualified_relation_name(relationId);
+	char *tableOwnerName = TableOwner(relationId);
+
+	appendStringInfo(ownerResetCommand,
+					 "ALTER TABLE %s OWNER TO %s",
+					 qualifiedRelationName,
+					 quote_identifier(tableOwnerName));
+
+	return ownerResetCommand->data;
+}
+
+
+/*
+ * ShardListInsertCommand generates a singe command that can be
+ * executed to replicate shard and shard placement metadata for the
+ * given shard intervals. The function assumes that each shard has a
+ * single placement, and asserts this information.
+ */
+List *
+ShardListInsertCommand(List *shardIntervalList)
+{
+	List *commandList = NIL;
+	ListCell *shardIntervalCell = NULL;
+	StringInfo insertPlacementCommand = makeStringInfo();
+	StringInfo insertShardCommand = makeStringInfo();
+	int shardCount = list_length(shardIntervalList);
+	int processedShardCount = 0;
+	int processedShardPlacementCount = 0;
+
+	/* if there are no shards, return empty list */
+	if (shardCount == 0)
+	{
+		return commandList;
+	}
+
+	/* generate the shard placement query without any values yet */
+	appendStringInfo(insertPlacementCommand,
+					 "INSERT INTO pg_dist_shard_placement "
+					 "(shardid, shardstate, shardlength,"
+					 " nodename, nodeport) "
+					 "VALUES ");
+
+	/* add placements to insertPlacementCommand */
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		uint64 shardId = shardInterval->shardId;
+
+		List *shardPlacementList = FinalizedShardPlacementList(shardId);
+		ShardPlacement *placement = NULL;
+
+		/* the function only handles single placement per shard */
+		Assert(list_length(shardPlacementList) == 1);
+
+		placement = (ShardPlacement *) linitial(shardPlacementList);
+
+		appendStringInfo(insertPlacementCommand,
+						 "(%lu, 1, %lu, %s, %d)",
+						 shardId,
+						 placement->shardLength,
+						 quote_literal_cstr(placement->nodeName),
+						 placement->nodePort);
+
+		processedShardPlacementCount++;
+		if (processedShardPlacementCount != shardCount)
+		{
+			appendStringInfo(insertPlacementCommand, ",");
+		}
+	}
+
+	/* add the command to the list that we'll return */
+	commandList = lappend(commandList, insertPlacementCommand->data);
+
+	/* now, generate the shard query without any values yet */
+	appendStringInfo(insertShardCommand,
+					 "INSERT INTO pg_dist_shard "
+					 "(logicalrelid, shardid, shardstorage,"
+					 " shardminvalue, shardmaxvalue) "
+					 "VALUES ");
+
+	/* now add shards to insertShardCommand */
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		uint64 shardId = shardInterval->shardId;
+		Oid distributedRelationId = shardInterval->relationId;
+		char *qualifiedRelationName = generate_qualified_relation_name(
+			distributedRelationId);
+
+		/* TODO: consider other types */
```

Comment:
I think we can remove this todo. It's part of a much wider issue of supporting range- and append-distributed tables.